tico/experimental/quantization/passes/insert_quantize_on_dtype_mismatch.py

```diff
@@ -0,0 +1,288 @@
+    - To resolve the issue, we insert quantize operators not to violate circle's type inference logic.
+    """
+
+    def __init__(self):
+        super().__init__()
+
+    def call(self, exported_program: ExportedProgram) -> PassResult:
+        logger = logging.getLogger(__name__)
+
+        graph_module = exported_program.graph_module
+        graph: torch.fx.Graph = graph_module.graph
+
+        def _insert_quantize_op_before(node, inp):
+            qparam: QuantParam = node.meta[QPARAM_KEY]
+            assert qparam.scale is not None
+            assert qparam.zero_point is not None
+            scale = qparam.scale[0]
+            zerop = qparam.zero_point[0]
+            min_ = qparam.min[0] if qparam.min is not None else 0
+            max_ = qparam.max[0] if qparam.max is not None else 0
+            dtype = getattr(torch, qparam.dtype)
+
+            with graph.inserting_before(node):
+                q_args = (inp, scale, zerop, min_, max_, dtype)
+                quantize = graph.call_function(
+                    torch.ops.quantized_decomposed.quantize_per_tensor.default,
+                    args=q_args,
+                )
+                set_new_meta_val(quantize)
+
+            node.replace_input_with(inp, quantize)
+
+            return quantize
+
+        def _insert_quantize_op_after(node):
+            qparam: QuantParam = node.meta[QPARAM_KEY]
+            assert qparam.scale is not None
+            assert qparam.zero_point is not None
+            scale = qparam.scale[0]
+            zerop = qparam.zero_point[0]
+            min_ = qparam.min[0] if qparam.min is not None else 0
+            max_ = qparam.max[0] if qparam.max is not None else 0
+            dtype = getattr(torch, qparam.dtype)
+            with graph.inserting_after(node):
+                q_args = (node, scale, zerop, min_, max_, dtype)
+                quantize = graph.call_function(
+                    torch.ops.quantized_decomposed.quantize_per_tensor.default,
+                    args=q_args,
+                )
+
+            node.replace_all_uses_with(quantize, propagate_meta=True)
+            quantize.replace_input_with(quantize, node)
+
+            return quantize
+
+        for node in graph.nodes:
+            if node.op != "call_function":
+                continue
+            if node.target == torch.ops.aten.linear.default:
+                lin_args = LinearArgs(*node.args, **node.kwargs)
+                inp = lin_args.input
+
+                if QPARAM_KEY not in inp.meta:
+                    continue
+
+                if QPARAM_KEY not in node.meta:
+                    continue
+
+                if _dtype(inp) == _dtype(node):
+                    continue
+
+                if _dtype(inp) == "uint8" and _dtype(node) == "int16":
+                    quantize = _insert_quantize_op_after(node)
+
+                    quantize.meta[QPARAM_KEY] = copy.deepcopy(node.meta[QPARAM_KEY])
```

Comment:
```suggestion
                    quantize.meta[QPARAM_KEY] = node.meta[QPARAM_KEY]
```

Reply:
This can cause sharing of qparam. Copy would be safer.